File: python-benchmarks/benchmarks/bson_benchmark.py
"""
BSON Benchmark
===============
Benchmark for the BSON (Binary JSON) format.
Uses the bson module from pymongo.

BSON is the binary format used internally by MongoDB.
It supports additional data types such as datetime, binary, and ObjectId.
"""

import bson
from .base_benchmark import BaseBenchmark


class BsonBenchmark(BaseBenchmark):
    """Benchmark for BSON serialization/deserialization."""

    @property
    def format_name(self) -> str:
        return "BSON"

    def serialize(self, data: dict) -> bytes:
        """Serialize dict to BSON bytes."""
        return bson.encode(data)

    def deserialize(self, payload: bytes) -> dict:
        """Deserialize BSON bytes to dict."""
        return bson.decode(payload)

File: go-benchmarks/benchmarks/bson_benchmark.go
package benchmarks

import (
	"go.mongodb.org/mongo-driver/bson"
)

// BSONBenchmark benchmarks the BSON (Binary JSON) format.
// BSON is the binary format used internally by MongoDB.
type BSONBenchmark struct{}

func (b *BSONBenchmark) FormatName() string {
	return "BSON"
}

func (b *BSONBenchmark) Serialize(data map[string]interface{}) ([]byte, error) {
	return bson.Marshal(data)
}

func (b *BSONBenchmark) Deserialize(payload []byte) (map[string]interface{}, error) {
	var result map[string]interface{}
	err := bson.Unmarshal(payload, &result)
	return result, err
}
